src/System.Net.Security/src/System/Net/Security/SecureChannel.cs

```diff
@@ -624,15 +624,26 @@ private bool AcquireClientCredentials(ref byte[] thumbPrint)
         //
         // Acquire Server Side Certificate information and set it on the class.
         //
-        private bool AcquireServerCredentials(ref byte[] thumbPrint)
+        private bool AcquireServerCredentials(ref byte[] thumbPrint, byte[] clientHello)
         {
             if (NetEventSource.IsEnabled)
                 NetEventSource.Enter(this);
 
             X509Certificate localCertificate = null;
             bool cachedCred = false;
 
-            if (_sslAuthenticationOptions.CertSelectionDelegate != null)
+            if (_sslAuthenticationOptions.ServerCertSelectionDelegate != null)
+            {
+                string serverIdentity = SniHelper.GetServerName(clientHello);
```

Comment:
SChannel really makes you parse the hello manually?  I'd much rather see the whole parser dropped if the implementations will give back the string(s).

Reply:
Schannel actually has similar API to this one but is not available on Windows 7. Other reason we chose parsing is that certificate is passed to security context constructor where OpenSSL way to get SNI is to set a callback after construction and callback is called when handshake function is called. This is doable but fairly risky. Having parsing allows us to do other things easier (i.e. make cipher suites be available in the callback) and have uniform code across all platforms. Since we already got parsing it doesn't make much sense to do different things on other platforms